MoltenVK/MoltenVK/Commands/MVKCommandEncoderState.mm

```diff
@@ -54,34 +54,43 @@
 												  uint32_t firstViewport,
 												  bool isSettingDynamically) {
 
-	bool mustSetDynamically = _cmdEncoder->supportsDynamicState(VK_DYNAMIC_STATE_VIEWPORT);
-
 	uint32_t maxViewports = _cmdEncoder->getDevice()->_pProperties->limits.maxViewports;
-	if ((mustSetDynamically == isSettingDynamically) && mtlViewports.size() > 0 &&
-		(firstViewport + mtlViewports.size() <= maxViewports) &&
-		(firstViewport < maxViewports)) {
+	if ((firstViewport + mtlViewports.size() > maxViewports) ||
+		(firstViewport <= maxViewports) ||
+		(isSettingDynamically && mtlViewports.size() == 0))
+		return;
 
-		if (firstViewport + mtlViewports.size() > _mtlViewports.size()) {
-			_mtlViewports.resize(firstViewport + mtlViewports.size());
-		}
+	auto& usingMTLViewports = isSettingDynamically ? _mtlDynamicViewports : _mtlViewports;
 
-		std::copy(mtlViewports.begin(), mtlViewports.end(), _mtlViewports.begin() + firstViewport);
-		markDirty();
+	if (firstViewport + mtlViewports.size() > usingMTLViewports.size()) {
+		usingMTLViewports.resize(firstViewport + mtlViewports.size());
 	}
+
+	bool mustSetDynamically = _cmdEncoder->supportsDynamicState(VK_DYNAMIC_STATE_VIEWPORT);
+
+	if (isSettingDynamically ||
+		(!isSettingDynamically && !mustSetDynamically && mtlViewports.size() > 0))
+		std::copy(mtlViewports.begin(), mtlViewports.end(), usingMTLViewports.begin() + firstViewport);
+	else if (!isSettingDynamically && (mustSetDynamically || mtlViewports.size() == 0))
+		usingMTLViewports.clear();
```

Comment:
If I'm parsing the boolean logic of the `if` & `else if` conditions, I think it can be simplified:

- The `!isSettingDynamically &&` phrase in both the `if` and `else if` are redundant, since `!isSettingDynamically` is guaranteed to be `true` at that point in each case, otherwise the initial `isSettingDynamically` would have triggered the `if` block.
- The `(mustSetDynamically || mtlViewports.size() == 0)` in the `else if` is redundant as it is the logical negation of the  `!mustSetDynamically && mtlViewports.size() > 0` in the `if` (ie. (A || B) == !(!A & !B) ).

So could this just be simplified to:

    if (isSettingDynamically || (!mustSetDynamically && mtlViewports.size() > 0))
    ...
    else
    ...	

Please confirm my logic! And if I've made a mistake and the two branches are not complimentary...then what is the effect of neither of the `if` of `else if` blocks executing and then `markDirty()` being called?

Reply:
I believe you're right. The logical tree here is such that one of those two blocks will always be called, which requires that indeed it can be simplified to `if-else`. Will check and commit.